Sources/WendyAgent/Services/ContainerMonitor.swift

```diff
@@ -0,0 +1,264 @@
+                        self = .onFailure(maxRetries: maxRetries)
+                    } else {
+                        self = .onFailure(maxRetries: 5)  // Default to 5 retries
+                    }
+                } else {
+                    self = .no  // Default policy
+                }
+            }
+        }
+    }
+
+    private var containerStates: [String: ContainerState] = [:]
+
+    /// Start monitoring containers
+    func startMonitoring() {
+        guard !monitoring else { return }
+        monitoring = true
+
+        monitorTask = Task {
+            logger.info("Starting container monitor service")
+
+            // Small delay to let containerd fully start after boot
+            try? await Task.sleep(nanoseconds: 2_000_000_000)
+
+            while monitoring {
+                do {
+                    try await checkContainers()
+                    try await Task.sleep(nanoseconds: 15_000_000_000)  // Check every 15 seconds
+                } catch {
+                    logger.error("Error in container monitor", metadata: ["error": "\(error)"])
+                    try? await Task.sleep(nanoseconds: 30_000_000_000)  // Wait 30 seconds on error
+                }
+            }
+
+            logger.info("Container monitor service stopped")
+        }
+    }
+
+    /// Stop monitoring containers
+    func stopMonitoring() {
+        monitoring = false
+        monitorTask?.cancel()
+        monitorTask = nil
+    }
+
+    /// Mark a container as explicitly stopped (won't be auto-restarted)
+    func markContainerStopped(_ appName: String) {
+        if containerStates[appName] != nil {
+            containerStates[appName]?.explicitlyStopped = true
+            logger.info("Marked container as explicitly stopped", metadata: ["container": "\(appName)"])
+        }
+    }
+
+    /// Mark a container as running (can be auto-restarted again)
+    func markContainerStarted(_ appName: String) {
+        if containerStates[appName] != nil {
+            containerStates[appName]?.explicitlyStopped = false
+            containerStates[appName]?.failureCount = 0
+            logger.info("Marked container as started", metadata: ["container": "\(appName)"])
+        }
+    }
+
+    /// Check container status and restart if needed
+    private func checkContainers() async throws {
+        // Use containerd API to get containers and their labels
+        let (containers, tasks) = try await Containerd.withClient { client in
+            let tasks = try await client.listTasks()
+            let containers = try await client.listContainers()
+            return (containers, tasks)
+        }
+
+        var wendyContainers = 0
+        for container in containers {
+            // Check if this is a Wendy-managed container
+            guard container.labels["sh.wendy/app.version"] != nil else {
+                // skip non-Wendy containers
+                continue
```

Comment:
Don't we want all containers to restart? Including development builds

Reply:
All containers created by Wendy should have the above label, even development containers. This is to ignore other non-wendy containers that may have been created by the user or another process on the device. However, this could be removed if you think it's worth it to include other containers in this logic.